### DISCUSSION TITLE
When using nested patternProperties regex not works

### DISCUSSION THREAD

**Discussion by jeffersongreco** [json-schema-org/community#626] [Q&A]
When I use nested patternProperties, in the key the regex works but the error message is the default one and not the one I defined in the scheme and in the subkey the error follows what is defined but the regex does not work, it reports an error even though it is correct.
I'm using VSCode.

//color-semantic.schema.json
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "$schema": {
      "type": "string"
    }
  },
  "patternProperties": {
    "^[a-z0-9_]+$": {
      "$ref": "#/definitions/colorReference",
      "errorMessage": "Invalid property name. Use snake_case."
    }
  },
  "additionalProperties": false,
  "definitions": {
    "kebabCase": {
      "type": "string",
      "pattern": "^[a-z0-9-]+$",
      "errorMessage": "Invalid property name. Use kebab-case."
    },
    "colorReference": {
      "oneOf": [
        {
          "$ref": "#/definitions/kebabCase"
        },
        {
          "type": "object",
          "patternProperties": {
            "^[a-z0-9_]+$": {
              "$ref": "#/definitions/kebabCase",
              "errorMessage": "Invalid property name. Use snake_case."
            }
          }
        }
      ],
      "additionalProperties": false
    }
  }
}
```

//color-semantic.tokens.json
```json
{
  "$schema": "../src/types/color-semantic.schema.json",
  "surface-": { // Error: Property surface- is not allowed.
    "minus_1": "neutral-transparent-30", // Error: Invalid property name. Use snake_case.
    "0": "neutral-opaque-0",
    "1": "neutral-opaque-30",
    "2": "neutral-opaque-60"
  }
}
```

**Comment by gregsdennis** [comment]
`errorMessage` isn't a JSON Schema feature.  It might be something specific to the implementation you've been using.

JSON Schema, per the specification, will just ignore that property.

**Reply by gregsdennis** [reply]
You may also want to read through [this blog post](https://json-schema.org/blog/posts/interpreting-output), which explains why error messages might be generated for valid instances when the schema uses keywords like `oneOf`.

I suspect that VSCode is likely not handling this properly.

Regardless, this is something to raise with them.  We control the spec, but we don't have any say on any particular implementations.

**Comment by jdesrosiers** [comment]
I can't speak for proprietary VSCode functionality, but I did notice a bug in your schema that could be the cause of the problem

```json
  "oneOf": [
    {
      "$ref": "#/definitions/kebabCase"
    },
    {
      "type": "object",
      "patternProperties": {
        "^[a-z0-9_]+$": {
          "$ref": "#/definitions/kebabCase",
          "errorMessage": "Invalid property name. Use snake_case."
        }
      }
    }
  ],
  "additionalProperties": false
```

The scope of `additionalProperties` is only the keywords in the local subschema. The `patternProperties` keyword in this case is in a different subschema and `additionalProperties` doesn't know it exists. So, your schema is effectively, `{ "additionalProperties": false }`, which means the only JSON instance that will validate is the empty object (`{}`). If you move the `additionalProperties` keyword so it's at the same level as `patternProperties`, I suspect that everything will work as you expect.

There's a keyword called `unevaluatedProperties` that is very similar to `additionalProperties` except that it can take into account keywords in at the same level or below it in the JSON tree. If you had used that keyword, that would have been a valid schema that would evaluate as you expect. The only problem is that VSCode doesn't support this keyword. Fortunately for you, it doesn't appear that you actually need it. Moving `additionalProperties` to the right place seem like the right solution in this case.

**Reply by MrJarnould** [reply]
@jeffersongreco Note that `unevaluatedProperties` was introduced in 2019-09. So if you would like to use it, you'll have to update your JSON Schema since you're currently using draft 07.
